subprojects/logging/src/test/groovy/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandlerTest.groovy

```diff
@@ -16,42 +16,235 @@
 
 package org.gradle.internal.featurelifecycle
 
+import org.gradle.internal.logging.CollectingTestOutputEventListener
 import org.gradle.internal.logging.ConfigureLogging
-import org.gradle.internal.logging.TestOutputEventListener
 import org.gradle.util.SetSystemProperties
 import org.gradle.util.TextUtil
 import org.junit.Rule
 import spock.lang.Specification
+import spock.lang.Subject
+import spock.lang.Unroll
 
+@Subject(LoggingDeprecatedFeatureHandler)
 class LoggingDeprecatedFeatureHandlerTest extends Specification {
-    final outputEventListener = new TestOutputEventListener()
-    @Rule final ConfigureLogging logging = new ConfigureLogging(outputEventListener)
-    @Rule SetSystemProperties systemProperties = new SetSystemProperties()
+    final outputEventListener = new CollectingTestOutputEventListener()
+    @Rule
+    final ConfigureLogging logging = new ConfigureLogging(outputEventListener)
+    @Rule
+    SetSystemProperties systemProperties = new SetSystemProperties()
     final locationReporter = Mock(UsageLocationReporter)
     final handler = new LoggingDeprecatedFeatureHandler(locationReporter)
 
     def "logs each deprecation warning once only"() {
         when:
-        handler.deprecatedFeatureUsed(new DeprecatedFeatureUsage("feature1", LoggingDeprecatedFeatureHandlerTest))
-        handler.deprecatedFeatureUsed(new DeprecatedFeatureUsage("feature2", LoggingDeprecatedFeatureHandlerTest))
-        handler.deprecatedFeatureUsed(new DeprecatedFeatureUsage("feature2", LoggingDeprecatedFeatureHandlerTest))
+        handler.deprecatedFeatureUsed(new DeprecatedFeatureUsage("feature1", []))
+        handler.deprecatedFeatureUsed(new DeprecatedFeatureUsage("feature2", []))
+        handler.deprecatedFeatureUsed(new DeprecatedFeatureUsage("feature2", []))
 
         then:
-        outputEventListener.toString() == '[WARN feature1][WARN feature2]'
+        def events = outputEventListener.events
+        events.size() == 2
+
+        and:
+        events[0].message == 'feature1'
+        events[1].message == 'feature2'
     }
 
-    def "location reporter can prepend text"() {
-        def usage = new DeprecatedFeatureUsage("feature", LoggingDeprecatedFeatureHandlerTest)
+    @Unroll
+    def 'logs fake call Gradle script element first and #deprecationTracePropertyName=#deprecationTraceProperty'() {
+        System.setProperty(deprecationTracePropertyName, '' + deprecationTraceProperty)
```

Comment:
Is any special handling necessary to make `SetSystemProperties` work?
Because the `@Rule` is already applied to the spec.